File: frontend/src/types/conversation.ts
/**
 * Conversation Intelligence Types - Meeting summaries, call analysis
 */

// Enums
export type ConversationType =
  | "meeting"
  | "call"
  | "email_thread"
  | "chat"
  | "other";

export type SentimentType = "positive" | "neutral" | "negative" | "mixed";

// Main types
export interface ConversationAnalysis {
  id: string;
  contact_id: string | null;
  deal_id: string | null;
  activity_id: string | null;
  type: ConversationType;
  title: string;
  occurred_at: string;
  duration_minutes: number | null;
  participants: string[];
  summary: string;
  key_points: string[];
  action_items: ActionItem[];
  decisions_made: string[];
  questions_raised: string[];
  follow_up_required: boolean;
  sentiment: SentimentType;
  sentiment_score: number;
  topics: string[];
  keywords: string[];
  mentioned_people: string[];
  mentioned_companies: string[];
  created_at: string;
  analyzed_at: string;
  model_version: string;
}

export interface ActionItem {
  description: string;
  assignee: string | null;
  due_date: string | null;
  priority: "low" | "medium" | "high";
  completed: boolean;
}

export interface ConversationSummary {
  total_conversations: number;
  recent_summary: string | null;
  last_conversation: string | null;
  common_topics: string[];
  sentiment_trend: string;
  average_sentiment_score: number;
  pending_action_items: ActionItem[];
}

// Request types
export interface AnalyzeConversationRequest {
  type: ConversationType;
  title: string;
  transcript?: string;
  notes?: string;
  contact_id?: string;
  deal_id?: string;
  activity_id?: string;
  occurred_at?: string;
  duration_minutes?: number;
  participants?: string[];
}

export interface QuickSummarizeRequest {
  text: string;
  context?: string;
}

export interface QuickSummaryResponse {
  summary: string;
  key_points: string[];
  action_items: string[];
}

// Helper functions
export function getSentimentColor(sentiment: SentimentType): string {
  switch (sentiment) {
    case "positive":
      return "text-green-600";
    case "negative":
      return "text-red-600";
    case "mixed":
      return "text-yellow-600";
    default:
      return "text-gray-600";
  }
}

export function getSentimentBgColor(sentiment: SentimentType): string {
  switch (sentiment) {
    case "positive":
      return "bg-green-100 text-green-800";
    case "negative":
      return "bg-red-100 text-red-800";
    case "mixed":
      return "bg-yellow-100 text-yellow-800";
    default:
      return "bg-gray-100 text-gray-800";
  }
}

export function getConversationTypeLabel(type: ConversationType): string {
  switch (type) {
    case "meeting":
      return "Meeting";
    case "call":
      return "Call";
    case "email_thread":
      return "Email Thread";
    case "chat":
      return "Chat";
    default:
      return "Other";
  }
}

export function formatDuration(minutes: number | null): string {
  if (!minutes) return "";
  if (minutes < 60) return `${minutes}m`;
  const hours = Math.floor(minutes / 60);
  const mins = minutes % 60;
  return mins > 0 ? `${hours}h ${mins}m` : `${hours}h`;
}

File: backend/app/models/conversation.py
"""Models for Conversation Intelligence - meeting summaries, call analysis."""

from datetime import datetime
from enum import Enum
from typing import Optional
from uuid import UUID, uuid4

from pydantic import BaseModel as PydanticBaseModel
from sqlalchemy import Column
from sqlalchemy.dialects.postgresql import JSONB
from sqlmodel import Field, SQLModel


class ConversationType(str, Enum):
    """Type of conversation."""
    MEETING = "meeting"
    CALL = "call"
    EMAIL_THREAD = "email_thread"
    CHAT = "chat"
    OTHER = "other"


class SentimentType(str, Enum):
    """Overall sentiment of conversation."""
    POSITIVE = "positive"
    NEUTRAL = "neutral"
    NEGATIVE = "negative"
    MIXED = "mixed"


class ConversationAnalysis(SQLModel, table=True):
    """Database model for conversation analysis."""

    __tablename__ = "conversation_analyses"

    id: UUID = Field(default_factory=uuid4, primary_key=True)

    # Associated entities
    contact_id: Optional[UUID] = Field(default=None, foreign_key="contacts.id", index=True)
    deal_id: Optional[UUID] = Field(default=None, foreign_key="deals.id", index=True)
    activity_id: Optional[UUID] = Field(default=None, foreign_key="activities.id", index=True)

    # Conversation metadata
    type: ConversationType = Field(default=ConversationType.OTHER)
    title: str = Field(max_length=200)
    occurred_at: datetime = Field(default_factory=datetime.utcnow)
    duration_minutes: Optional[int] = None
    participants: list = Field(default_factory=list, sa_column=Column(JSONB))

    # Original content
    transcript: Optional[str] = None  # Full transcript if available
    notes: Optional[str] = None  # User notes

    # AI Analysis
    summary: str  # AI-generated summary
    key_points: list = Field(default_factory=list, sa_column=Column(JSONB))
    action_items: list = Field(default_factory=list, sa_column=Column(JSONB))
    decisions_made: list = Field(default_factory=list, sa_column=Column(JSONB))
    questions_raised: list = Field(default_factory=list, sa_column=Column(JSONB))
    follow_up_required: bool = Field(default=False)

    # Sentiment analysis
    sentiment: SentimentType = Field(default=SentimentType.NEUTRAL)
    sentiment_score: float = Field(default=0.0, ge=-1.0, le=1.0)
    sentiment_details: dict = Field(default_factory=dict, sa_column=Column(JSONB))

    # Extracted entities
    mentioned_people: list = Field(default_factory=list, sa_column=Column(JSONB))
    mentioned_companies: list = Field(default_factory=list, sa_column=Column(JSONB))
    mentioned_products: list = Field(default_factory=list, sa_column=Column(JSONB))
    mentioned_dates: list = Field(default_factory=list, sa_column=Column(JSONB))
    mentioned_amounts: list = Field(default_factory=list, sa_column=Column(JSONB))

    # Topics and keywords
    topics: list = Field(default_factory=list, sa_column=Column(JSONB))
    keywords: list = Field(default_factory=list, sa_column=Column(JSONB))

    # Timestamps
    created_at: datetime = Field(default_factory=datetime.utcnow)
    analyzed_at: datetime = Field(default_factory=datetime.utcnow)

    # AI metadata
    model_version: str = Field(default="llama3.1")


class ActionItem(PydanticBaseModel):
    """An action item extracted from a conversation."""
    description: str
    assignee: Optional[str] = None
    due_date: Optional[str] = None
    priority: str = "medium"
    completed: bool = False


# Response models
class ConversationAnalysisResponse(PydanticBaseModel):
    """API response for conversation analysis."""
    id: UUID
    contact_id: Optional[UUID]
    deal_id: Optional[UUID]
    activity_id: Optional[UUID]
    type: ConversationType
    title: str
    occurred_at: datetime
    duration_minutes: Optional[int]
    participants: list
    summary: str
    key_points: list
    action_items: list
    decisions_made: list
    questions_raised: list
    follow_up_required: bool
    sentiment: SentimentType
    sentiment_score: float
    topics: list
    keywords: list
    mentioned_people: list
    mentioned_companies: list
    created_at: datetime
    analyzed_at: datetime
    model_version: str

    class Config:
        from_attributes = True


class AnalyzeConversationRequest(PydanticBaseModel):
    """Request to analyze a conversation."""
    type: ConversationType = ConversationType.OTHER
    title: str
    transcript: Optional[str] = None
    notes: Optional[str] = None
    contact_id: Optional[UUID] = None
    deal_id: Optional[UUID] = None
    activity_id: Optional[UUID] = None
    occurred_at: Optional[datetime] = None
    duration_minutes: Optional[int] = None
    participants: list[str] = []


class QuickSummarizeRequest(PydanticBaseModel):
    """Request for quick summarization without full analysis."""
    text: str
    context: Optional[str] = None
